### PR TITLE
Add readout error mitigation to qat

### PR DESCRIPTION
**Pull request by jfriel-oqc**
Adds some simple readout mitigation in the form of full matrix (to become sparse) and a constant overhead linear approximation. M3 to be added once we have a hybrid runtime

same as PR on bitbucket, just ported over

Not all the tests were passing (note that were on bitbucket). Will investigate this that comes from this PR or the port over

### CODE REVIEW

**Review by chemix-lunacy — CHANGES_REQUESTED** [review]
The return dictionary and regex variable splicing will need to be fixed before this goes in.

File: src/QAT/qat/purr/compiler/execution.py [hunk 1]
```diff
@@ -225,6 +229,32 @@ def squash_binary(value):
...
+            if isinstance(inst, PostProcessingReadoutMitigation)
+        ]
+        if not readout_instructions:
+            return results
+
+        if len(results) > 1:
+            # TODO: add support for multiple registers
+            raise ValueError(
+                "Cannot have multiple registers in conjunction with readout error mitigation."
+            )
```
Comment (hamidelmaazouz) [inline src/QAT/qat/purr/compiler/execution.py:248]: this check can precede earlier processing to raise early (unless need to use `readout_instructions` when supporting multiple registers)

**Review by hamidelmaazouz — CHANGES_REQUESTED** [review]

**Comment by jfriel-oqc** [comment]
Reopening this PR, myself and @jdavid-oqc will address these comments this week

File: src/QAT/qat/purr/compiler/execution.py [hunk 2]
```diff
@@ -225,6 +229,32 @@ def squash_binary(value):
 
         return results
 
+    def _apply_error_mitigation(self, results, qfile: "QatFile"):
+        """
+        Process any software-driven results transformation, such as taking a raw waveform result
+        and turning it into a bit, or something else.
+        """
+        readout_instructions = [
```
[the remark this answers is not in the document]
Comment (jfriel-oqc) [reply src/QAT/qat/purr/compiler/execution.py:237]: yeah, that's true, I thought this was an easier API for execution via internal builder, but agree I think it's easier to execute with config

**Comment by jfriel-oqc** [comment]
Declining as many changes ported over in https://github.com/oqc-community/qat/pull/70

**Pull request closed without merging**